rfcs/20200424-sparse-domain-isolation.md

```diff
@@ -0,0 +1,649 @@
+- Look up the newest values from `dynamic_embedding.Variable` and update them to memory held by `TrainableWrap->tensor()` before each iteration start.
+
+
+#### python/ops/resource_variable_ops.py:
+
+```python
+
+class TrainableWrap(ResourceVariable):
+  """
+  This class is a trainable warp of dynamic embedding,
+  and the key role is recording the map relationship of params and ids.
+  inheriting from the ResourceVariable make it trainable.
+  """
+  def __init__(self, params, ids, *args, **kwargs):
+    """Creates an empty `TrainableWrap` object.
+
+    Creates a group of tables placed on devices,
+    the type of its keys and values are specified by key_dtype
+    and value_dtype, respectively.
+
+    Args:
+      params: A dynamic_embedding.Variable instance.
+      ids: a tensor with any shape as same dtype of params.key_dtype.
+      ```other parameters is similar to tf.nn.embedding_lookup.```
+    Returns:
+      A `TrainableWrap` object which is a subclass of ResourceVariable.
+    """
+    self.params = params
+    self.ids = ids
+    self.prefetch_values = self.params.lookup(self.ids)
+
+    super(TrainableWrap, self).__init__(*args, **kwargs)
+
+  """A python variable from an existing handle."""
+  def _read_variable_op(self):
+
+    # The `assign_variable_op` will be called to get before read opertion.
+    # `prefetch_values` is a lookup-like operation of `TrainableWrap`.
+
+    if hasattr(self, "prefetch_values"):
+      with ops.control_dependencies([
+          gen_resource_variable_ops.assign_variable_op(
+            self._handle, self.prefetch_values,  
+          name="AssignBeforeReadVariable")]):
+        result = gen_resource_variable_ops.read_variable_op(self._handle,
+                                                            self._dtype)
+    else:
+      result = gen_resource_variable_ops.read_variable_op(self._handle,
+                                                          self._dtype)
+    _maybe_set_handle_data(self._dtype, self._handle, result)
+
+
+  def update_op(self):
+    return self.params.upsert(self.ids, self._read_variable_op())
+```
+
+The scheme will **reuse** the `Optimizer._resource_apply_{dense,sparse_duplicate_indices}}` API to train `resource_variable_ops.TrainableWrap` naturally after changing a little common code in `training/optimizer.py`:
+
+- Create the slots of `TrainableWrap` when primary `Variable` is an instance of `TrainableWrap`.
+- Write back the trained values of primary and slots to hash tables(`dynamic_embedding.Variable`) when `Optimizer._resource_apply_{dense,sparse_duplicate_indices}` finished.
+
+#### The pseudo-code is below: 
+
+```python
+@tf_export("dynamic_embedding.create_slots")
+def create_slots(primary,
+                 init,
+                 slot_name,
+                 op_name):
+  """Helper function for creating a slot variable for statefull optimizers."""
+  _params_var, _params_ids = _IDS_OF_RESOURCEVARIABLE_MAPPER_.get(primary)
```

Comment:
What is `_IDS_OF_RESOURCEVARIABLE_MAPPER_ `? 

It looks like your slot variable is created for every step since it depends on the `ids` (while there is not such requirement in the original optimizers). Is that correct?

Reply:
> What is `_IDS_OF_RESOURCEVARIABLE_MAPPER_ `?
> 
> It looks like your slot variable is created for every step since it depends on the `ids` (while there is not such requirement in the original optimizers). Is that correct?

I will cancel the `_IDS_OF_RESOURCEVARIABLE_MAPPER_ ` by reusing the `Optimizer.slots`